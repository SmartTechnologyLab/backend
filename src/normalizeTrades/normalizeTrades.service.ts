import { Injectable } from '@nestjs/common';
import { GrouppedTrades, ITrade } from 'src/report/types';
import { groupBy } from 'ramda';
import { StockExchange } from './constants';
import { IFreedomFinanceTrade } from '../report/types/freedomFinance';

@Injectable()
export class NormalizeTradesService {
  constructor() {}

  private MAP_STOCK_EXCHANGE_TO_TRADE_TYPE = {
    [StockExchange.FREEDOM_FINANCE]:
      this.normalizedFreedomFinanceTrades.bind(this),
  };

  getNormalizedTrades(
    stockExchange: StockExchange,
    trades: unknown[],
  ): ITrade[] {
    return this.MAP_STOCK_EXCHANGE_TO_TRADE_TYPE[stockExchange](
      trades as IFreedomFinanceTrade[],
    );
  }

  normalizedFreedomFinanceTrades(
    trades: (IFreedomFinanceTrade | ITrade)[],
  ): ITrade[] {
    const conformingTrades: ITrade[] = trades.filter(this.isITrade);
    const nonConformingTrades = trades.filter(
      (trade) => !this.isITrade(trade),
    ) as IFreedomFinanceTrade[];

    const mappedTrades: ITrade[] = nonConformingTrades.map((trade) => ({
      ticker: trade.instr_nm?.split('.').at(0),
      price: trade.p,
      commission: trade.commission,
      operation: trade.operation,
      quantity: trade.q,
      date: trade.date,
      currency: trade.curr_c,
    }));

    return [...conformingTrades, ...mappedTrades];
  }

  isITrade(trade: unknown): trade is ITrade {
    return trade.hasOwnProperty('ticker');
  }
}
