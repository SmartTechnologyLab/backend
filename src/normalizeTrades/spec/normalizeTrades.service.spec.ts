import { NormalizeTradesService } from '../normalizeTrades.service';
import { StockExchange } from '../constants';
import {
  freedomFinanceTrades,
  freedomFinanceTradesWithPreviousPeriods,
} from './__fixtures__/freedomFinance';

describe('Normalize Trades Service', () => {
  let normalizeTradesService: NormalizeTradesService;

  beforeEach(() => {
    normalizeTradesService = new NormalizeTradesService();
  });

  describe('normalize trades returns trades converted to interface ITrade', () => {
    it('freedom finance trades', () => {
      const normalizedTrades = normalizeTradesService.getNormalizedTrades(
        StockExchange.FREEDOM_FINANCE,
        freedomFinanceTrades,
      );

      expect(normalizedTrades).toMatchSnapshot();
    });
  });

  it('normalized freedom finance trades return trades converted to interface ITrade for IFreedomFinanceTrade', () => {
    const normalizedFreedomFinanceTrades =
      normalizeTradesService.normalizedFreedomFinanceTrades(
        freedomFinanceTradesWithPreviousPeriods,
      );

    expect(normalizedFreedomFinanceTrades).toMatchSnapshot();
  });

  describe('validate if object is already consist of ITrade interface', () => {
    it('returns true if object contains property of ticker', () => {
      const isITrade = normalizeTradesService.isITrade({
        ticker: 'AAPL',
        price: 100,
        commission: 0,
        operation: 'Buy',
        quantity: 10,
        date: '2021-01-01',
        currency: 'USD',
      });

      expect(isITrade).toBeTruthy();
    });

    it('returns false if object contains property of ticker', () => {
      const isITrade = normalizeTradesService.isITrade({
        isnt_nm: 'AAPL',
        price: 100,
        commission: 0,
        operation: 'Buy',
        quantity: 10,
        date: '2021-01-01',
        currency: 'USD',
      });

      expect(isITrade).toBeFalsy();
    });
  });
});
