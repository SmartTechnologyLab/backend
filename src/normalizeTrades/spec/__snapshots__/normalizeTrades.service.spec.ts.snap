// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalize Trades Service getReportByStockExchange returns trades by stock exchange returns trades for freedom finance stock exchange 1`] = `
[
  {
    "commission": 2.31,
    "curr_c": "USD",
    "date": "2021-06-07 11:48:21",
    "instr_nm": "FIPO.US",
    "operation": "buy",
    "p": 35.53,
    "q": 13,
  },
  {
    "commission": 2.31,
    "curr_c": "USD",
    "date": "2021-06-07 11:48:21",
    "instr_nm": "FIPO.US",
    "operation": "buy",
    "p": 35.53,
    "q": 13,
  },
  {
    "commission": 2.31,
    "curr_c": "USD",
    "date": "2021-06-07 11:48:21",
    "instr_nm": "FIPO.US",
    "operation": "sell",
    "p": 35.53,
    "q": 26,
  },
  {
    "commission": 7.75,
    "curr_c": "USD",
    "date": "2021-06-10 14:06:17",
    "instr_nm": "MNDY_IPO.US",
    "operation": "buy",
    "p": 155,
    "q": 1,
  },
  {
    "commission": 7.75,
    "curr_c": "USD",
    "date": "2021-07-10 14:06:17",
    "instr_nm": "MNDY_IPO.BLOCKED",
    "operation": "buy",
    "p": 156,
    "q": 3,
  },
  {
    "commission": 7.75,
    "curr_c": "sell",
    "date": "2021-10-10 14:06:17",
    "instr_nm": "MNDY_IPO",
    "operation": "sell",
    "p": 155,
    "q": 3,
  },
  {
    "commission": 7.75,
    "curr_c": "sell",
    "date": "2021-11-10 14:06:17",
    "instr_nm": "MNDY_IPO",
    "operation": "sell",
    "p": 155,
    "q": 1,
  },
  {
    "commission": 1.75,
    "curr_c": "USD",
    "date": "2021-06-10 16:30:00",
    "instr_nm": "WISH.US",
    "operation": "buy",
    "p": 11.06,
    "q": 10,
  },
  {
    "commission": 1.75,
    "curr_c": "USD",
    "date": "2021-06-10 16:30:00",
    "instr_nm": "WISH.US",
    "operation": "sell",
    "p": 11.06,
    "q": 10,
  },
  {
    "commission": 1.73,
    "curr_c": "USD",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "RIDE",
    "operation": "buy",
    "p": 10.65,
    "q": 10,
  },
  {
    "commission": 1.73,
    "curr_c": "sell",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "RIDE",
    "operation": "sell",
    "p": 10.65,
    "q": 12,
  },
  {
    "commission": 1.73,
    "curr_c": "sell",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "RIDE",
    "operation": "buy",
    "p": 10.65,
    "q": 2,
  },
  {
    "commission": 1.73,
    "curr_c": "sell",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "GOOGL",
    "operation": "sell",
    "p": 10.65,
    "q": 10,
  },
  {
    "commission": 1.73,
    "curr_c": "sell",
    "date": "2021-06-11 18:41:12",
    "instr_nm": "GOOGL",
    "operation": "buy",
    "p": 10.65,
    "q": 10,
  },
  {
    "commission": 1.73,
    "curr_c": "USD",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "APPL.US",
    "operation": "buy",
    "p": 10.65,
    "q": 10,
  },
]
`;

exports[`Normalize Trades Service groupTradesByTicker groups trades by ticker 1`] = `
{
  "APPL": [
    {
      "commission": 1.73,
      "currency": "USD",
      "date": "2021-06-10 18:41:12",
      "operation": "buy",
      "price": 10.65,
      "quantity": 10,
      "ticker": "APPL",
    },
  ],
  "FIPO": [
    {
      "commission": 2.31,
      "currency": "USD",
      "date": "2021-06-07 11:48:21",
      "operation": "buy",
      "price": 35.53,
      "quantity": 13,
      "ticker": "FIPO",
    },
    {
      "commission": 2.31,
      "currency": "USD",
      "date": "2021-06-07 11:48:21",
      "operation": "buy",
      "price": 35.53,
      "quantity": 13,
      "ticker": "FIPO",
    },
    {
      "commission": 2.31,
      "currency": "USD",
      "date": "2021-06-07 11:48:21",
      "operation": "sell",
      "price": 35.53,
      "quantity": 26,
      "ticker": "FIPO",
    },
  ],
  "GOOGL": [
    {
      "commission": 1.73,
      "currency": "sell",
      "date": "2021-06-10 18:41:12",
      "operation": "sell",
      "price": 10.65,
      "quantity": 10,
      "ticker": "GOOGL",
    },
    {
      "commission": 1.73,
      "currency": "sell",
      "date": "2021-06-11 18:41:12",
      "operation": "buy",
      "price": 10.65,
      "quantity": 10,
      "ticker": "GOOGL",
    },
  ],
  "MNDY_IPO": [
    {
      "commission": 7.75,
      "currency": "USD",
      "date": "2021-06-10 14:06:17",
      "operation": "buy",
      "price": 155,
      "quantity": 1,
      "ticker": "MNDY_IPO",
    },
    {
      "commission": 7.75,
      "currency": "USD",
      "date": "2021-07-10 14:06:17",
      "operation": "buy",
      "price": 156,
      "quantity": 3,
      "ticker": "MNDY_IPO",
    },
    {
      "commission": 7.75,
      "currency": "sell",
      "date": "2021-10-10 14:06:17",
      "operation": "sell",
      "price": 155,
      "quantity": 3,
      "ticker": "MNDY_IPO",
    },
    {
      "commission": 7.75,
      "currency": "sell",
      "date": "2021-11-10 14:06:17",
      "operation": "sell",
      "price": 155,
      "quantity": 1,
      "ticker": "MNDY_IPO",
    },
  ],
  "RIDE": [
    {
      "commission": 1.73,
      "currency": "USD",
      "date": "2021-06-10 18:41:12",
      "operation": "buy",
      "price": 10.65,
      "quantity": 10,
      "ticker": "RIDE",
    },
    {
      "commission": 1.73,
      "currency": "sell",
      "date": "2021-06-10 18:41:12",
      "operation": "sell",
      "price": 10.65,
      "quantity": 12,
      "ticker": "RIDE",
    },
    {
      "commission": 1.73,
      "currency": "sell",
      "date": "2021-06-10 18:41:12",
      "operation": "buy",
      "price": 10.65,
      "quantity": 2,
      "ticker": "RIDE",
    },
  ],
  "WISH": [
    {
      "commission": 1.75,
      "currency": "USD",
      "date": "2021-06-10 16:30:00",
      "operation": "buy",
      "price": 11.06,
      "quantity": 10,
      "ticker": "WISH",
    },
    {
      "commission": 1.75,
      "currency": "USD",
      "date": "2021-06-10 16:30:00",
      "operation": "sell",
      "price": 11.06,
      "quantity": 10,
      "ticker": "WISH",
    },
  ],
}
`;

exports[`Normalize Trades Service normalize trades returns trades converted to interface ITrade freedom finance trades 1`] = `
[
  {
    "commission": 2.31,
    "currency": "USD",
    "date": "2021-06-07 11:48:21",
    "operation": "buy",
    "price": 35.53,
    "quantity": 13,
    "ticker": "FIPO",
  },
  {
    "commission": 2.31,
    "currency": "USD",
    "date": "2021-06-07 11:48:21",
    "operation": "buy",
    "price": 35.53,
    "quantity": 13,
    "ticker": "FIPO",
  },
  {
    "commission": 2.31,
    "currency": "USD",
    "date": "2021-06-07 11:48:21",
    "operation": "sell",
    "price": 35.53,
    "quantity": 26,
    "ticker": "FIPO",
  },
  {
    "commission": 7.75,
    "currency": "USD",
    "date": "2021-06-10 14:06:17",
    "operation": "buy",
    "price": 155,
    "quantity": 1,
    "ticker": "MNDY_IPO",
  },
  {
    "commission": 7.75,
    "currency": "USD",
    "date": "2021-07-10 14:06:17",
    "operation": "buy",
    "price": 156,
    "quantity": 3,
    "ticker": "MNDY_IPO",
  },
  {
    "commission": 7.75,
    "currency": "sell",
    "date": "2021-10-10 14:06:17",
    "operation": "sell",
    "price": 155,
    "quantity": 3,
    "ticker": "MNDY_IPO",
  },
  {
    "commission": 7.75,
    "currency": "sell",
    "date": "2021-11-10 14:06:17",
    "operation": "sell",
    "price": 155,
    "quantity": 1,
    "ticker": "MNDY_IPO",
  },
  {
    "commission": 1.75,
    "currency": "USD",
    "date": "2021-06-10 16:30:00",
    "operation": "buy",
    "price": 11.06,
    "quantity": 10,
    "ticker": "WISH",
  },
  {
    "commission": 1.75,
    "currency": "USD",
    "date": "2021-06-10 16:30:00",
    "operation": "sell",
    "price": 11.06,
    "quantity": 10,
    "ticker": "WISH",
  },
  {
    "commission": 1.73,
    "currency": "USD",
    "date": "2021-06-10 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 10,
    "ticker": "RIDE",
  },
  {
    "commission": 1.73,
    "currency": "sell",
    "date": "2021-06-10 18:41:12",
    "operation": "sell",
    "price": 10.65,
    "quantity": 12,
    "ticker": "RIDE",
  },
  {
    "commission": 1.73,
    "currency": "sell",
    "date": "2021-06-10 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 2,
    "ticker": "RIDE",
  },
  {
    "commission": 1.73,
    "currency": "sell",
    "date": "2021-06-10 18:41:12",
    "operation": "sell",
    "price": 10.65,
    "quantity": 10,
    "ticker": "GOOGL",
  },
  {
    "commission": 1.73,
    "currency": "sell",
    "date": "2021-06-11 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 10,
    "ticker": "GOOGL",
  },
  {
    "commission": 1.73,
    "currency": "USD",
    "date": "2021-06-10 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 10,
    "ticker": "APPL",
  },
]
`;

exports[`Normalize Trades Service normalized freedom finance trades return trades converted to interface ITrade for IFreedomFinanceTrade 1`] = `
[
  {
    "commission": 1.73,
    "currency": "sell",
    "date": "2021-06-11 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 10,
    "ticker": "GOOGL",
  },
  {
    "commission": 1.73,
    "currency": "USD",
    "date": "2021-06-10 18:41:12",
    "operation": "buy",
    "price": 10.65,
    "quantity": 10,
    "ticker": "APPL",
  },
]
`;
